File: GramPositive.php
<?php
//version 1.0
function tokenize($text) {
    // Take text block, separate into sentences, remove punctuation, 
    // and then separate into words
    $text = strtolower($text); // Make text lowercase

    // Normalize punctuation for splitting and remove commas
    $text = str_replace(["!", "?"], ".", $text);
    $text = str_replace(",", "", $text);

    $textParts = explode(".", $text);
    $sentences = [];

    foreach ($textParts as $sent) {
        $st = explode(" ", $sent);
        
        // Remove empty strings from the array
        $st = array_filter($st, function($word) {
            return $word !== "";
        });

        // Reset array keys so it behaves like a clean Python list
        $st = array_values($st);

        if (!empty($st)) {
            $sentences[] = $st;
        }
    }

    return $sentences;} 



function process_text($json_file_path, $text, $word_match = false) {
    if (!file_exists($json_file_path)) { 
        return ["error" => "Model file missing"]; 
    }
    $json_string = file_get_contents($json_file_path);
    $model = json_decode($json_string, true);
    
    $word_index = $model["index"];
    $matrix_model = $model["matrix"];
    $default_label = $model["default"];
    $scores = [];
    
    $sentences = tokenize($text);
    $wordcount = 0;
    
    // Set your maximum allowed character edits (1 is best for typos/leet speak)
    $lev_threshold = 1; 
    
    foreach ($sentences as $sent) {
        $prec = "*43$#00"; // set precedent to *43$#00 if there is no precedent
        
        foreach ($sent as $word) {
            // If word not in word_index
            if (!isset($word_index[$word])) {
                if ($word_match === true) {
                    $closest_word = null;
                    $shortest_dist = $lev_threshold + 1;
                    
                    // Replaces Python's lst_about by checking the index keys natively
                    foreach ($word_index as $valid_word => $id) {
                        $dist = levenshtein($word, $valid_word);
                        
                        if ($dist < $shortest_dist) {
                            $closest_word = $valid_word;
                            $shortest_dist = $dist;
                        }
                        
                        // Early exit if we find a perfect 1-edit match to save CPU cycles
                        if ($shortest_dist === 1) {
                            break;
                        }
                    }
                    
                    // If a close match was found within our threshold, use it
                    if ($closest_word !== null && $shortest_dist <= $lev_threshold) {
                        $word = $closest_word;
                    }
                } else {
                    continue; // Equivalent to Python's 'pass' when word_match is False
                }
            }
            
            // If word in word_index
            if (isset($word_index[$word])) {
                $wordcount++; // count the words so that an average score can be calculated
                
                $value = $word_index[$word];
                $prec_value = isset($word_index[$prec]) ? $word_index[$prec] : "";
                $pattern = "{$prec_value}:{$value}";
                
                $prec = $word; // the prec only gets updated if the word exists
                
                // Get the pattern scores but only if the pattern exists in the model
                if (isset($matrix_model[$pattern])) {
                    $labels = $matrix_model[$pattern];
                } else {
                    // fallback for unknown patterns
                    $labels = [$default_label => 1];
                }
                
                // Update the scores with the current labels
                foreach ($labels as $label => $label_value) {
                    if (isset($scores[$label])) {
                        $scores[$label] += $label_value;
                    } else {
                        $scores[$label] = $label_value;
                    }
                }
            }
        }
    }
    
    // Calculate average to avoid getting higher scores for longer sentences
    if ($wordcount !== 0) {
        $avg_scores = [];
        foreach ($scores as $label => $score) {
            $avg_scores[$label] = $score / $wordcount;
        }
        $scores = $avg_scores;
    }
    
    return $scores;
}

function classify($json_file_path, $text, $word_match = false) {
    // 1. Get the scores from the processing function
    $results = process_text($json_file_path, $text, $word_match);
    
    $verdict = "none";
    $highest = 0;
    
    // 2. Loop through results to find the key with the highest score
    foreach ($results as $label => $score) {
        if ($score > $highest) {
            $highest = $score;
            $verdict = $label;
        }
    }
    
    return $verdict;
}
?>
<!-- ////////////////////////////////////   Demonstration   //////////////////////////////////////////////////// -->
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Model Tester</title>
    <style>
        body { font-family: Arial, sans-serif; margin: 40px; background: #f9f9f9; }
        .container { max-width: 600px; background: white; padding: 20px; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }
        textarea { width: 100%; height: 100px; margin-bottom: 15px; padding: 10px; box-sizing: border-box; }
        input[type="submit"] { background: #007BFF; color: white; border: none; padding: 10px 20px; border-radius: 4px; cursor: pointer; }
        input[type="submit"]:hover { background: #0056b3; }
        .result { margin-top: 20px; padding: 15px; border-left: 5px solid #28a745; background: #e9f7ef; }
    </style>
</head>
<body>

<div class="container">
    <h2>Text Classification Model Test</h2>
    
    <form method="post" action="">
        <label for="text_input">Enter text to classify:</label>
        <textarea name="text_input" id="text_input" required><?php echo isset($_POST['text_input']) ? htmlspecialchars($_POST['text_input']) : ''; ?></textarea>
        <input type="submit" value="Analyze Text">
    </form>

    <?php
    // --- 1. RUN THE MODEL IF FORM IS SUBMITTED ---
    if ($_SERVER["REQUEST_METHOD"] == "POST" && !empty($_POST['text_input'])) {
        
        $text_to_test = $_POST['text_input'];
        
        // run the classify function to return the result
        $prediction = classify("model.json", $text_to_test, false);
        
        // Display the result
        echo "<div class='result'>";
        echo "<strong>Classification Result:</strong> " . htmlspecialchars($prediction);
        echo "</div>";
    }
    ?>
</div>

File: GramPositive.py
__version__="1.0"
import json
from string_comparator import lst_about
class GramPositive:
    def __init__(self,model:dict=None):
        self.model=model
    def tokenize(self,text):
        '''take text block, separate into sentences, remove punctuation, 
        and then separate into words'''
        text=text.lower()#make text lowercase

        text=text.replace("!",".").replace("?",".").replace(",","")#normalize punctuation for splitting and remove commas
        text=text.split(".")
        sentences=[]
        for sent in text:
            st=sent.split(" ")
            st = [word for word in st if word != ""]
            if st!=[]:
                sentences.append(st)

        return sentences

    def train_gram_model(self,input_list:dict,default_label="normal"):
        '''trains an n-gram model file by looking for common patterns in a dictionary of
        prelabled text strings
        input structure: {text:lebel,text:lebel,text:lebel} 
        model structure: model={"index":{0:0},"matrix":{"0:0":label,0:0":label}}'''
        if self.model:#load model from input
            word_index=self.model["index"]
            matrix_model=self.model["matrix"]
            default_label=self.model["default"]
            current_value=len(word_index)+1
        else:  #generate new model
            word_index={"*43$#00":0}
            matrix_model={}
            current_value=1#the last value in the index

        for text_block in input_list: #iterate though the whole training data and tokenize and process each example
            label=input_list[text_block]
            sentences=self.tokenize(text_block)#tokenize
            for sent in sentences:#iterate and process through each sentence
                prec="*43$#00" #set precedent to *43$#00 if there is no precedent (less chances for incidents)
                for word in sent: #take each word, put it in the model and find patterns
                    if word in word_index:#get the value from the index if already there 
                        value=word_index.get(word)
                    else:#add the word in index if not there and update the next value
                        word_index[word]=current_value
                        value=current_value
                        current_value+=1
                    pattern=f"{word_index[prec]}:{value}" #generate the pattern based on the precedent word
                    prec=word
                    if pattern not in matrix_model: #pair the lables to each pattern and add them to matrix
                        matrix_model[pattern]={label:1}# each label will get a score based on how often it is flagges in each text later allowing quantization
                    else:
                        labels=matrix_model[pattern]#update each label or add a new one for existent patterns
                        if label in labels:
                            labels[label]+=1
                        else:
                            labels[label]=1
                        matrix_model[pattern]=labels
        #package the model
        self.model={"default":default_label,
        "index":word_index,
        "matrix":matrix_model}
        return self.model
    def process_text(self,text,word_match=False):
        word_index=self.model["index"]
        matrix_model=self.model["matrix"]
        default_label=self.model["default"]
        scores={}
        sentences=self.tokenize(text)
        wordcount=0
        if word_match==True:
            all_words=[i for i in word_index]
        for sent in sentences:
            prec="*43$#00" #set precedent to *43$#00 if there is no precedent (less chances for incidents)
            for word in sent:
                if word not in word_index:
                    if word_match==True:
                        word=lst_about(word,all_words)
                    else:
                        pass
                if word in word_index:
                    wordcount+=1#count the words so that an average score can be calculated
                    value=word_index.get(word)
                    pattern=f"{word_index[prec]}:{value}"
                    prec=word #the prec only gets updated if the word exists, thus the function will comletely skip ant unknown words
                    if pattern in matrix_model: #get the patern scores but only if the pattern or word exists in the model
                        labels=matrix_model[pattern]
                        
                    else:#fallback for unknown patterns.
                        labels={default_label:1}
                    for label in labels: #update the scores with the current labels
                        if label in scores:
                            scores[label] += labels[label]
                        else:
                            scores[label] = labels[label]

        #calculate average to avoid getting higher scores for longer sentences
        if wordcount!=0:
            avg_scores={}
            for i in scores:
                avg_scores[i]=scores[i]/wordcount
            scores=avg_scores
        return scores
    def classify(self,text,word_match=False):
        results=self.process_text(text,word_match)
        verdict="none"
        highest=0
        for i in results:
            if results[i]>highest:
                verdict=i
        return verdict
    def quantize(self,level=90):
        '''trims the patterns that have more than level percent of all possible labels'''
        if self.model:
            matrix_model = self.model["matrix"]
            label_list=[]
            for pattern in matrix_model:#get all labels to determine the threshold for removal
                for i in matrix_model[pattern]:
                    if i not in label_list:
                        label_list.append(i)
            threshold=round((level*len(label_list))/100) #calculate treshold as [level] % of all labels
            quant_model={}
            for pattern in matrix_model:#remove all patterns over treshold
                if len(matrix_model[pattern])<=threshold:
                    quant_model[pattern]=matrix_model[pattern]
            self.model["matrix"]=quant_model
            return self.model
        else:
            print("please make sure there is a model to qunatize")
    def load_model(self,path="model.json"):
        with open(path,"r") as f:
            self.model=json.load(f)
    def save_model(self,path="model.json"):
        with open(path,"w") as f:
            json.dump(self.model,f)